### PR TITLE
creates memcached in a new subnet group

### PR DESCRIPTION
**Pull request by scriptnull**
https://github.com/devops-recipes/prov_aws_memcached_terraform/issues/1

Currently provisioning of memcached fails in devops-recipes - https://app.shippable.com/github/devops-recipes/jobs/prov_aws_memcached_tf/builds/5b31d22034d4ed070035d776/console with the following error.

```
Error: Error applying plan:

1 error(s) occurred:

* aws_elasticache_cluster.memcachedCluster: 1 error(s) occurred:

* aws_elasticache_cluster.memcachedCluster: error creating Elasticache Cache Cluster: InvalidParameterValue: The account does not have any default subnets.
	status code: 400, request id: [REDACTED]

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```

Implementing the simplest access pattern ( memcached and ec2 in same VPC ) defined in [AWS documentation](https://docs.aws.amazon.com/AmazonElastiCache/latest/mem-ug/elasticache-vpc-accessing.html#elasticache-vpc-accessing-same-vpc)

### CODE REVIEW

**Comment by rageshkrishna** [comment]
I'm sure that the AWS account used for devops-recipes must already have a VPC and subnet available for use, even if they are not defaults.

@avinci owns that environment, so I'd like him to chime in on what's the right configuration to be used for this.

**Comment by avinci** [comment]
We already have a vpc. Why can’t we use that?

**Comment by avinci** [comment]
we should not do this. We already have a VPC.

**Comment by scriptnull** [comment]
Ok. I will update this PR to use a subnet created in our already existing VPC ( in https://github.com/devops-recipes/prov_aws_vpc_terraform ). I thought of checking if it works in this configuration and to use our VPC after confirming it.

**Comment by scriptnull** [comment]
Changed the PR to implement the simplest access pattern mentioned in AWS documentation
- https://docs.aws.amazon.com/AmazonElastiCache/latest/mem-ug/elasticache-vpc-accessing.html#elasticache-vpc-accessing-same-vpc
- https://docs.aws.amazon.com/AmazonElastiCache/latest/mem-ug/VPCs.EC.html ( Refer [diagram](https://docs.aws.amazon.com/AmazonElastiCache/latest/mem-ug/images/vpc-overview-diagram.png) for overview )

File: memcached.tf [hunk 1]
```diff
@@ -4,13 +4,28 @@ provider "aws" {
   region      = "${var.aws_region}"
 }
 
+resource "aws_elasticache_subnet_group" "demo_memcached_subnet_group" {
```
Comment (avinci) [inline memcached.tf:7]: what is this for? why do we need a subnet group?

File: memcached.tf [hunk 2]
```diff
@@ -17,7 +17,7 @@ resource "aws_elasticache_cluster" "memcachedCluster" {
   parameter_group_name = "default.memcached1.4"
   port                 = 11211
   subnet_group_name    = "${aws_elasticache_subnet_group.demo_memcached_subnet_group.name}"
```
Comment (avinci) [inline memcached.tf]: still not sure what this is

File: memcached.tf (hunk 1 above)
Comment (scriptnull) [reply memcached.tf:7]: This resource is for helping Amazon Elasticache to identify the subnets ( in a VPC ) to be used for spinning up the cache nodes.

The resource [aws_elasticache_cluster](https://www.terraform.io/docs/providers/aws/r/elasticache_cluster.html) accepts only [aws_elasticache_subnet_group](https://www.terraform.io/docs/providers/aws/r/elasticache_subnet_group.html) and there is no argument like `subnet_id`. So, we are forced to create this resource, even if we need to mention only one subnet in our subnet group.

Comment (avinci) [reply memcached.tf:7]: why is this happening on public groups? then what was the point of the private group?

Comment (scriptnull) [reply memcached.tf:7]: Do you mean `vpc_public_sn_id` being used in this subnet group?

Our current implementation looks like this reference picture from [AWS docs](https://docs.aws.amazon.com/AmazonElastiCache/latest/mem-ug/VPCs.EC.html)

Image Source: https://docs.aws.amazon.com/AmazonElastiCache/latest/mem-ug/VPCs.EC.html

Both the memcached and EC2 instances are spin up in the same subnet, but we use two different  security groups to make sure that EC2 instances are available to the outside world and memcached cluster is available only to the machines inside the subnet.

**Pull request merged**